src/database.ts

```diff
@@ -72,42 +76,93 @@ export class Database<D = unknown> {
    * @param statements A set of Statement objects to batch and submit.
    * @param opts Additional options to control execution.
    * @returns An array of run results.
    */
+  //    Note: if we want this package to mirror the D1 package in a way that
+  //    enables compatability with packages built to exend D1, then the return type
+  //    here will potentially affect if/how those packages work.
+  //    D1-ORM is a good example: https://github.com/Interactions-as-a-Service/d1-orm/
   async batch<T = D>(
     statements: Statement[],
     opts: Signal = {}
-  ): Promise<Array<Result<T>>> {
+    // reads returns an Array with legnth equal to the number of batched statements,
+    // everything else a single result wrapped in an Array for backward compatability.
+    // TODO: In order to work around the Validator API not returning all of the tableIds
+    //       and continuing to work in a backward compatable way, it seems that we have to
+    //       make this type `any` :(
+    //       We should attempt to fix this when the Validator API update happens, or on the
+    //       next major version.
+  ): Promise<any> {
     try {
+      const start = performance.now();
+      // If the statement types are "create" and the statement contains more than one
+      // query (separated by semi-colon) then the sqlparser with throw an Error.
       const normalized = await Promise.all(
         statements.map(async (stmt) => await normalize(stmt.toString()))
       );
+
       const type: string | null = normalized
         .map((stmt) => stmt.type)
         .reduce((a, b): any => (a === b ? a : null));
       if (type == null) {
         throw new Error(
-          "statement error: batch must contain uniform types (e.g., CREATE, INSERT, SELECT, etc)"
+          "statement error: batch must contain uniform types (i.e. one of: create, write, read, acl)"
         );
       }
+
+      // "read" statement types are the simple case, we just do each of the queries
+      // and return an Array of the query results.
       if (type === "read") {
         return await Promise.all(
           statements.map(async (stmt) => await stmt.all<T>(undefined, opts))
         );
-      } else {
-        // Mutating queries are sent as a single query to the smart contract
-        const sql = statements.map((stmt) => stmt.toString()).join(";");
-        const result = await this.prepare(sql).all<T>(undefined, opts);
-        return [result];
       }
+
+      // For "create" statement types, each statement must be a single create sql query
+      if (type === "create") {
+        const receipt = await checkWait(
+          this.config,
+          await execCreateMany(
+            this.config,
+            statements.map((stmt) => stmt.toString())
+          )
+        );
+
+        // TODO: wrapping in an Array is required for back compat, consider changing this for next major
+        return [wrapResult(receipt, performance.now() - start)];
```

Comment:
Since this is a single transaction, we _could_ return an Object instead of an Array, but I'm not sure if that breaks D1 compatibility.

Reply:
Think it's good to keep compatability.